Added CONTRIBUTING.md as requested in an issue. (#360)
novelrt/NovelRT: CONTRIBUTING.md

# CONTRIBUTING GUIDELINES

> NovelRT is a community-driven project. Contribution is aslways welcome, and much appreciated.

Reviewer: ```suggestion
NovelRT is a community-driven project. Contribution is always welcome, and much appreciated.
```

Just a suggestion here as always appears to be mistyped

It is also how it'll be developed.

> First, please take a moment to review our [code of conduct](CODE_OF_CONDUCT.md).

Reviewer: ```suggestion
First, please take a moment to review our [code of conduct](CODE_OF_CONDUCT.md). We expect that all project participants will adhere to it, and any actions taken that go against the Code of Conduct will not be tolerated.
```

Good first line here! Just added a suggestion as we want to convey that actions against the CoC have zero tolerance in our eyes.

> These guidelines are an attempt at better addressing the huge amount of pending
> issues and pull requests. Please read them closely.

Reviewer: ```suggestion
These guidelines are an attempt at better addressing the incoming issues and pull requests. 
Please read them closely, and feel free to ask any questions if needed.
```

Just a suggestion here as the softer language makes it sound generally more inviting 🙂

> ## Submitting Pull Requests
> 
> ### Getting started

Reviewer: ```suggestion
## Getting started

### Submitting Pull Requests
```

Suggestion: Reverse these? It seems to read easier this way (in the context of the flow of the document).

You might know the basics of 
[contributing on GitHub](https://help.github.com/articles/using-pull-requests) and have a fork
[properly set up](https://github.com/ohmyzsh/ohmyzsh/wiki/Contribution-Technical-Practices).

You can only create PRs on a dedicated branch.

If you create your own PR, please double check that everything is all right. Also please type what you fixed or added in the PR body,
[for instance](https://help.github.com/articles/closing-issues-via-commit-messages/)

## Volunteer

> Very nice!! :)

Reviewer: We will need a little more detail here on how contributors can get started 😅 apologies for not having the issue fully written out at the time you picked this up.

More specifically, we want to cover the following:

- How to Get in Touch
-- If they want to ask about a ticket, make a suggestion, show off how they're using NovelRT, if they need assistance with the project or even just chat, they can reach us at Discord 😄 
-- They can also open an issue or start a Github Discussion, too
-- Also, if they have additional (relevant) input on currently open issues, they can feel free to comment as well

- Looking for an Issue
-- Call out here that we do use `help wanted` and `good first issue` labels, but any issue that is unassigned can be picked up.
-- If the issue you're looking at is assigned but hasn't been worked on in a considerable amount of time, reach out to the maintainers about reassignment.

- Contributing to Development (once an Issue has been picked up)
-- This area should link back on how to build NovelRT (from the README)
-- It should also mention our Code Style - we use an automatic formatter `clang-format` to adhere to a codestyle similar to Microsoft's (https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/code-style-rule-options?view=vs-2017)
-- Also should mention that if they have trouble adhering to the code style, our CI process will produce a patch for them when needed so that they can apply it and make their code adherent.

- Creating a Pull Request
-- This should mention that PRs need to be directed to the `main` branch, but should start from a fork on a different branch.
-- You can specifically call out if they are a first-timer, they can look to the basic links that you provided here.
-- PRs should be tagged with an issue they resolve - PRs without issue need to be briefly discussed with the maintainers first before they will be accepted.
-- Once a PR has been submitted, one or more of the maintainers will review, comment, and/or merge based on if the pull request meets criteria or not.
-- It should be specifically called out that any pull requests created must have their creator agree to our Contributor License Agreement (CLA) before the pull request will be allowed to be merged.

If you'd like a better idea of how the flow of the guide should be, feel free to look here as a reference: https://reactjs.org/docs/how-to-contribute.html

Also, for this piece, you can tag me in comments or reach out to me however if you need additional follow-up or info to help you complete this!

> Thank you to [all the contributers!](https://github.com/novelrt/NovelRT/graphs/contributors)

Reviewer: ```suggestion
Finally, thank you to [all of our current and future contributors!](https://github.com/novelrt/NovelRT/graphs/contributors)
```

One last suggestion here - just tweaked it a bit and corrected 1 typo (I like the premise here though)

# CONTRIBUTING GUIDELINES

NovelRT is a community-driven project. Contribution is always welcome, and much appreciated.

> It is also how it'll be developed.

Reviewer 2: Unsure if this line is really neccessary, as the statement right before this kind of states this, I feel?

> First, please take a moment to review our [code of conduct](CODE_OF_CONDUCT.md). We expect that all project participants will adhere to it, and any actions taken that go against the Code of Conduct will not be tolerated.

Reviewer 2: I personally think this reads better:

```suggestion
First, please take a moment to review our [Code of Conduct](CODE_OF_CONDUCT.md). We expect that all project participants will adhere to it, and violations will not be tolerated.

You can forward violations or other concerns to admin@novelrt.dev or via a direct message to one of the maintainers on our [Discord community server](https://discord.novelrt.dev).
```

> These guidelines are an attempt at better addressing the incoming issues and pull requests. 
> Please read them closely, and feel free to ask any questions if needed.

Reviewer 2: @capnkenny gonna want you to check my suggestion here, I'm not 100% sure on my suggestion here but I do think the original could be improved.

```suggestion
These guidelines are an outline of the contribution process. Please read them closely, and feel free to ask any questions if needed.
```

Reviewer: I think it summarizes the phrasing well here @RubyNova

## Getting started
### Submitting Pull Requests

> You might know the basics of 
> [contributing on GitHub](https://help.github.com/articles/using-pull-requests) and have a fork
> [properly set up](https://github.com/ohmyzsh/ohmyzsh/wiki/Contribution-Technical-Practices).

Reviewer 2: I think we have two options here.

You can either do this:
```suggestion
You might know the basics of [contributing on GitHub](https://help.github.com/articles/using-pull-requests) and have a fork [properly set up](https://github.com/ohmyzsh/ohmyzsh/wiki/Contribution-Technical-Practices).
```

Or you can do this:
```suggestion
Before submitting a pull request (PR), please ensure you are familiar with the following: 
- [Contributing on GitHub](https://help.github.com/articles/using-pull-requests)
- [Properly setting up a fork](https://github.com/ohmyzsh/ohmyzsh/wiki/Contribution-Technical-Practices).
```

I personally prefer the latter, but I'm open to discuss it.

Reviewer: I'd agree to the latter as well

You can only create PRs on a dedicated branch.

> If you create your own PR, please double check that everything is all right. Also please type what you fixed or added in the PR body,
> [for instance](https://help.github.com/articles/closing-issues-via-commit-messages/)

Reviewer 2: I think the message would come across clearer like so:

```suggestion
When creating a PR, please provide an outline of what you fixed or added in the PR body with relevant GitHub keywords and the issue number(s) in question. You can find more information on this [here](https://help.github.com/articles/closing-issues- via-commit-messages/).

Please also ensure that if you add new functionality, and it is testable, that you also add the relevant unit tests, if none exist.

While any feedback that is given from the maintainer team or one of our other contributors should be assumed to be in good faith and for the betterment of the NovelRT project and its sister repositories, please forward any concerns to the relevant places outlined at the beginning of this document.
```

It's more wordy, but I feel this covers the message better. :)
cc @capnkenny thoughts here?

Reviewer: Its more concise and direct in this version, @RubyNova

Your PR should be directed to the ```main``` branch - but should start from another fork of this repo.
If you need help / information - please ask for help! Especially if you're a ```first-timer```. 

#### If you contribute to the development, remember to keep this in mind.
[How you can build NovelRT](README.md) :)

> We use a code formatter named ```clang-format```, so our codestyle is similar to Microsoft's. 

Reviewer 2: Just adding some missing context here. :)

```suggestion
We use the ```clang-format``` code formatter configured to the Microsoft code formatting style, and will be using ```clang-tidy``` in the future to enforce conventions such as casing. For now, though, please ensure that your code follows the common C# styles, as used in places such as the [.NET Runtime](https://github.com/dotnet/runtime) and many other projects.
```

> If you find it hard to adhere to the code style, our CI process will produce a patch for you when neededm so that you can apply it and make your code adherent.

Reviewer 2: Fixing a spelling error and tweaking the wording a little with this suggestion.

```suggestion
If you find it hard to adhere to the code style, our CI process will produce a patch for you when needed so that you can apply it and make your code consistent with the rest of the project.
```

To submit a PR, you must accept the Contributor License Agreement (CLA) before the PR can be merged.

### Submitting Issues 
#### If you have a problem, 
Please be so kind as to check if anyone else already have submitted an issue about the same topic.

If you find one, feel free to comment on it so we can know there are multiple people experiencing it.

If didn't find any issue matching yours, then feel free to open an issue.
When you are making an issue, please include these topics if possible:

    When the problem occurs
    What the correct behavior should be
    What the actual behavior is

#### If you are looking for an issue to resolve,
Please feel free to resolve issues with labels like ```help wanted``` or good ```first issue``` even if it isn't your first issue etc. 

If you find an assigned issue that haven't been worked on for some time, then you're allowed to reach out and ask for a reassignment. 

#### Of course, you can always also 
  - Open a GitHub discussion
  - Comment on issues
  - Etc. 

> ### You can always reach us on [Discord](https://camo.githubusercontent.com/0912384154686aae2ca109266fb033dea65bfc0a5200cb1175e86de06bf68e23/68747470733a2f2f646973636f72646170702e636f6d2f6170692f6775696c64732f3534333839383936383338303134353637352f7769646765742e706e673f7374796c653d62616e6e657232) if you want to ask about a ticket, make a suggestion, show off how you're using NovelRT, if you need assistance with the project or even just chat. :) 

Reviewer 2: Using our official URL to make life easier. :)

```suggestion
### You can always reach us on [Discord](https://discord.novelrt.dev) if you want to ask about a ticket, make a suggestion, show off how you're using NovelRT, if you need assistance with the project or even just chat. :) 
```
